application-task-manager-api/src/main/java/com/xwiki/taskmanager/internal/TaskProcessor.java

```diff
@@ -79,21 +93,24 @@
 
     /**
      * Extracts the existing Tasks from a given XDOM.
+     *
      * @param content the XDOM from which one desires to extract or check for the existence of Tasks.
-     * @return a list of found Tasks or an empty list if the XDOM didn't contain any valid task. Where a valid task
-     * is one that has an id.
+     * @param taskCounter a counter used to assign an ID to the macros that do not have one assigned.
+     * @return a list of found Tasks or an empty list if the XDOM didn't contain any valid task. Where a valid task is
+     *     one that has an id.
      */
-    public List<Task> extract(XDOM content)
+    public List<Task> extract(XDOM content, TaskCounter taskCounter)
```

Comment:
The ``TaskCounter`` component could be replaced by a ``TaskReferenceGenerator``, but I'm not sure you need to pass it to the constructor. It could also be injected, unless you want to be able to use different task reference generators.